### ISSUE TITLE
Shut down problem

### ISSUE THREAD

**Issue by quarksgluons** [EOSIO/eos#5194]
The nodeos shut down,I can find nothing except the out_of_range_exception.
There is about 5 hours without data updates, and then the probram shut down.
Similar problem links have been closed but no solutions have been given.
[https://github.com/EOSIO/eos/issues/3029](url)
[https://github.com/EOSIO/eos/issues/3071](url)

```
124168 2018-08-11T17:14:37.592 thread-0   producer_plugin.cpp:330       on_incoming_block    ] Received block fe897dfb2ba2cbfd... #10126000 @ 2018-08       -08T19:42:26.000 signed by teamgreymass [trxs: 6, lib: 10125677, conf: 0, latency: 250331596 ms]
124169 2018-08-11T17:14:37.600 thread-0   net_plugin.cpp:2155           operator()           ] Peer eosbp.buildteam.io:8532 closed connection
124170 2018-08-11T17:14:37.601 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to eosnode.fi:9888: Connection refus       ed
124171 2018-08-11T17:14:38.546 thread-0   net_plugin.cpp:2155           operator()           ] Peer eu-west-nl.eosamsterdam.net:9876 closed connectio       n
124172 2018-08-11T17:14:40.304 thread-0   net_plugin.cpp:2155           operator()           ] Peer eosapi.blockmatrix.network:13546 closed connectio       n
124173 2018-08-11T17:14:44.211 thread-0   net_plugin.cpp:2155           operator()           ] Peer eu1.eosdac.io:49876 closed connection
124174 2018-08-11T17:14:48.405 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to p2p-1.eosnetwork.io:9876: Connect       ion timed out
124175 2018-08-11T17:14:59.420 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to mainnet.eosarabia.org:3571: Opera       tion canceled
124176 2018-08-11T17:14:59.426 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to mainnet.eosoasis.io:9876: Connect       ion refused
124177 2018-08-11T17:14:59.427 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to mainnet.eos.ren:9376: Connection        refused
124178 2018-08-11T17:14:59.428 thread-0   net_plugin.cpp:1935           connect              ] host: 106.10.42.238 port: 9876
124179 2018-08-11T17:14:59.428 thread-0   net_plugin.cpp:1935           connect              ] host: 123.59.116.52 port: 49876
124180 2018-08-11T17:14:59.428 thread-0   net_plugin.cpp:1935           connect              ] host: 130.211.59.178 port: 9876
...
...
127886 2018-08-11T17:32:07.800 thread-0   net_plugin.cpp:1222           process_next_message ] e.to_detail_string(): 8 out_of_range_exception: Out of        Range
127887 read datastream of length 1 over by 31
127888     {"method":"read","len":1,"over":31}
127889     thread-0  datastream.cpp:6 throw_datastream_range_error
127890 error unpacking N2fc14static_variantIJN5eosio5chain13chain_id_typeEEEE
127891     {"type":"N2fc14static_variantIJN5eosio5chain13chain_id_typeEEEE"}
127892     thread-0  raw.hpp:616 unpack
127893 Error unpacking field chain_id
127894     {"field":"chain_id"}
127895     thread-0  raw.hpp:368 operator()
127896 error unpacking eosio::handshake_message
127897     {"type":"eosio::handshake_message"}
127898     thread-0  raw.hpp:616 unpack
127899 2018-08-11T17:32:07.800 thread-0   net_plugin.cpp:2153           operator()           ] Error reading message from seed2.greymass.com:9876: Ba       d file descriptor
127900 2018-08-11T17:32:26.068 thread-0   net_plugin.cpp:1949           operator()           ] Unable to resolve bp.eosmedi.com:9877: Host not found        (authoritative)
127901 2018-08-11T17:32:26.076 thread-0   net_plugin.cpp:1949           operator()           ] Unable to resolve eos-p2p.worbli.io:33981: Host not fo       und (authoritative)
127902 2018-08-11T17:32:26.077 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to dns1-p2p.oraclechain.io:49876: Co       nnection refused
127903 2018-08-11T17:32:26.453 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to br.eosrio.io:9876: Co
...
...
272893 2018-08-12T00:03:07.623 thread-0   net_plugin.cpp:1935           connect              ] host: peering.eosio.cr port: 1976
272894 2018-08-12T00:03:07.623 thread-0   net_plugin.cpp:1935           connect              ] host: peering.eosio.cr port: 5418
272895 2018-08-12T00:03:07.623 thread-0   net_plugin.cpp:1935           connect              ] host: peering1.mainnet.eosasia.one port: 80
272896 2018-08-12T00:03:07.623 thread-0   net_plugin.cpp:1935           connect              ] host: peering2.mainnet.eosasia.one port: 80
272897 2018-08-12T00:03:07.623 thread-0   net_plugin.cpp:1935           connect              ] host: pub0.eosys.io port: 6637
272898 2018-08-12T00:03:07.623 thread-0   net_plugin.cpp:1935           connect              ] host: pub1.eostheworld.io port: 9876
272899 2018-08-12T00:03:07.623 thread-0   net_plugin.cpp:1935           connect              ] host: pub1.eosys.io port: 6637
272900 2018-08-12T00:03:07.623 thread-0   net_plugin.cpp:1935           connect              ] host: pub2.eostheworld.io port: 9876
272901 2018-08-12T00:03:09.837 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to 123.59.116.52:49876: Connection r       efused
272902 2018-08-12T00:03:09.861 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to boot.eostitan.com:9876: Operation        already in progress
272903 2018-08-12T00:03:09.885 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to boot.eostitan.com:9876: Operation        canceled
272904 2018-08-12T00:03:09.903 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to 106.10.42.238:9876: Connection re       fused
272905 2018-08-12T00:03:09.957 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to pub1.eosys.io:6637: Connection re       fused
272906 2018-08-12T00:03:09.978 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to 178.49.174.48:9876: Connection re       fused
272907 2018-08-12T00:03:10.044 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to 34.252.209.121:5556: Connection r       efused
272908 2018-08-12T00:03:10.086 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to 159.65.214.150:9876: Connection r       efused
272909 2018-08-12T00:03:10.101 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to 18.234.6.119:80: Connection refus       ed
272910 2018-08-12T00:03:10.116 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to 195.43.95.98:9876: Connection ref       used
272911 2018-08-12T00:03:10.141 thread-0   net_plugin.cpp:2155           operator()           ] Peer 807534da.eosnodeone.io:19872 closed connection
272912 2018-08-12T00:03:10.146 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to 94.130.250.22:9806: Connection re       fused
272913 2018-08-12T00:03:10.147 thread-0   net_plugin.cpp:1977           operator()           ] connection failed to 54.203.121.17:19866: Connection r       efused
272914 2018-08-12T00:03:10.292 thread-0   net_plugin.cpp:2155           operator()           ] Peer api.eosuk.io:12000 closed connection
272915 2018-08-12T00:03:10.417 thread-0   net_plugin.cpp:2155           operator()           ] Peer 185.253.188.1:19877 closed connection
272916 2018-08-12T00:03:10.500 thread-0   net_plugin.cpp:2155           operator()           ] Peer 18.191.33.148:59876 closed connection
```

**Comment by jgiszczak** [comment]
You have a great many peers which are completely dead, which is not helping diagnose the problem.  Remove all peers logged with "Host not found" and "Connection refused".

Typically the problem you're encountering is one of the ports you are connected to is the API port, not the net_plugin port.  You'll need to discover which host and port is speaking the wrong protocol and remove it.

If you're not synching at all, you should update your peer list to something more recent.  It's possible you had only one valid peer, and its connection limit has been reached, so you're not able to maintain a connection anymore.  The more valid peers you have, the better your chances of maintaining synchronization.

**Comment by taokayan** [comment]
you don't need you connect to all the peers. 2-3 peers should be ok for sync up.

**Comment by quarksgluons** [comment]
It stopped again last night and prompted replay required When I restarted it this morning.

```
2018-08-16T17:43:13.534 thread-0   net_plugin.cpp:2155           operator()           ] Peer mainnet.eosnairobi.io:9876 closed connection
2018-08-16T17:43:13.534 thread-0   net_plugin.cpp:2155           operator()           ] Peer p2p.eosio.cr:1976 closed connection
2018-08-16T17:43:13.574 thread-0   net_plugin.cpp:2155           operator()           ] Peer eos-bp.bitfinex.com:9876 closed connection
2018-08-16T17:43:13.577 thread-0   net_plugin.cpp:2155           operator()           ] Peer node1.eosnewyork.io:6987 closed connection
2018-08-16T17:43:13.588 thread-0   net_plugin.cpp:2155           operator()           ] Peer api.tokenika.io:9876 closed connection
2018-08-16T17:43:13.927 thread-0   net_plugin.cpp:2155           operator()           ] Peer bp.eosbeijing.one:8080 closed connection
2018-08-16T17:43:13.932 thread-0   net_plugin.cpp:2155           operator()           ] Peer cpt.eosio.africa:9876 closed connection
2018-08-16T17:43:13.932 thread-0   net_plugin.cpp:2155           operator()           ] Peer eos.nodepacific.com:9876 closed connection
2018-08-16T17:43:13.932 thread-0   net_plugin.cpp:2155           operator()           ] Peer seed2.eos42.io:9876 closed connection
2018-08-16T17:43:16.513 thread-0   net_plugin.cpp:2155           operator()           ] Peer p2p.eosdetroit.io:3018 closed connection
```

```
2018-08-17T00:35:05.058 thread-0   chain_plugin.cpp:271          plugin_initialize    ] initializing chain plugin
2018-08-17T00:35:05.071 thread-0   chain_plugin.cpp:580          plugin_initialize    ] 13 St13runtime_error: database dirty flag set
rethrow database dirty flag set:
    {"what":"database dirty flag set"}
    thread-0  chain_plugin.cpp:580 plugin_initialize
2018-08-17T00:35:05.072 thread-0   main.cpp:118                  main                 ] database dirty flag set (likely due to unclean shutdown): replay required
2018-08-17T00:36:09.289 thread-0   chain_plugin.cpp:271          plugin_initialize    ] initializing chain plugin
2018-08-17T00:36:09.290 thread-0   chain_plugin.cpp:580          plugin_initialize    ] 13 St13runtime_error: database dirty flag set
rethrow database dirty flag set:
    {"what":"database dirty flag set"}
    thread-0  chain_plugin.cpp:580 plugin_initialize
2018-08-17T00:36:09.290 thread-0   main.cpp:118                  main                 ] database dirty flag set (likely due to unclean shutdown): replay required
```
The peers in config.ini:
```
p2p-peer-address = 130.211.59.178:9876
p2p-peer-address = 159.65.214.150:9876
p2p-peer-address = 18.179.100.214:19876
p2p-peer-address = 47.104.169.45:9876
p2p-peer-address = 807534da.eosnodeone.io:19872
p2p-peer-address = anvil.eosblocksmith.io:9876
p2p-peer-address = api.eos.wiki:39876
p2p-peer-address = api.eosuk.io:12000
p2p-peer-address = api.tokenika.io:9876
p2p-peer-address = api1.eosmetal.io:19877
p2p-peer-address = api2.eostribe.io:9115
p2p-peer-address = api-full1.eoseoul.io:9876
p2p-peer-address = api-full2.eoseoul.io:9876
p2p-peer-address = bp.antpool.com:443
p2p-peer-address = bp.cryptolions.io:9876
p2p-peer-address = bp.eosbeijing.one:8080
p2p-peer-address = bp.eosmedi.com:9877
p2p-peer-address = bp.superone.io:57576
p2p-peer-address = cpt.eosio.africa:9876
p2p-peer-address = dc1.eosemerge.io:9876
p2p-peer-address = eos.genesis-mining.com:19876
p2p-peer-address = eos.genesis-mining.com:19877
p2p-peer-address = eos.nodepacific.com:9876
p2p-peer-address = eos.staked.us:9870
p2p-peer-address = eosapi.blockmatrix.network:13546
p2p-peer-address = eosboot.chainrift.com:9876
p2p-peer-address = eos-bp.bitfinex.com:9876
p2p-peer-address = eosbp.buildteam.io:8532
p2p-peer-address = eosbp.kuna.io:9876
p2p-peer-address = eosbp-0.atticlab.net:9876
p2p-peer-address = eosdapp.oneeos.org:9000
p2p-peer-address = eos-mainnet-p2p.activeeos.com:9876
p2p-peer-address = eos-seed-de.privex.io:9876
p2p-peer-address = eu.eosdac.io:49876
p2p-peer-address = fn001.eossv.org:443
p2p-peer-address = fullnode.eoslaomao.com:443
p2p-peer-address = m.jeda.one:3322
p2p-peer-address = mainnet.eosamsterdam.net:9876
p2p-peer-address = mainnet.eoscalgary.io:5222
p2p-peer-address = mainnet.eoseco.com:10010
p2p-peer-address = mainnet.eosnairobi.io:9876
p2p-peer-address = mainnet.libertyblock.io:9800
p2p-peer-address = mars.fnp2p.eosbixin.com:443
p2p-peer-address = node.eosflare.io:1883
p2p-peer-address = node.eosio.lt:9878
p2p-peer-address = node.eosmeso.io:9876
p2p-peer-address = node.eosvenezuela.io:9875
p2p-peer-address = node1.blockgenesys.com:9876
p2p-peer-address = node1.bp.eosindex.io:9876
p2p-peer-address = node1.eoscannon.io:59876
p2p-peer-address = node1.eosnewyork.io:6987
p2p-peer-address = node1.eosvibes.io:9876
p2p-peer-address = node1.starteos.io:9876
p2p-peer-address = node1.zbeos.com:9876
p2p-peer-address = node2.liquideos.com:9876
p2p-peer-address = node2.starteos.io:9876
p2p-peer-address = node2.zbeos.com:9876
p2p-peer-address = node869-mainnet.eosauthority.com:9393
p2p-peer-address = p2p.bitmars.one:8080
p2p-peer-address = p2p.bp.fish:9786
p2p-peer-address = p2p.eos.bitspace.no:9876
p2p-peer-address = p2p.eos.blckchnd.com:19876
p2p-peer-address = p2p.eosargentina.io:5222
p2p-peer-address = p2p.eosdetroit.io:3018
p2p-peer-address = p2p.eosgermany.online:9876
p2p-peer-address = p2p.eosio.cr:1976
p2p-peer-address = p2p.eosio.cr:5418
p2p-peer-address = p2p.eossocal.io:9999
p2p-peer-address = p2p.eossweden.se:9876
p2p-peer-address = p2p.genereos.io:9876
p2p-peer-address = p2p.hkeos.com:9876
p2p-peer-address = p2p.libertyblock.io:9800
p2p-peer-address = p2p.mainnet.eospace.io:88
p2p-peer-address = p2p.meet.one:9876
p2p-peer-address = p2p.one.eosdublin.io:9876
p2p-peer-address = p2p.proxy1a.sheos.org:5556
p2p-peer-address = p2p2.eossweden.eu:9966
p2p-peer-address = peer.eosgeneva.io:9876
p2p-peer-address = peer.eosn.io:9876
p2p-peer-address = peer.franceos.fr:19876
p2p-peer-address = peer.main.alohaeos.com:9876
p2p-peer-address = peer.oraclechain.io:19876
p2p-peer-address = peer1.eosphere.io:9876
p2p-peer-address = peer2.eoshuobipool.com:18181
p2p-peer-address = peer2.eosphere.io:9876
p2p-peer-address = peer2.franceos.fr:19876
p2p-peer-address = peer2.mainnet.helloeos.com.cn:80
p2p-peer-address = peering.mainnet.eoscanada.com:9876
p2p-peer-address = peering1.mainnet.eosasia.one:80
p2p-peer-address = pub0.eosys.io:6637
p2p-peer-address = publicnode.cypherglass.com:9876
p2p-peer-address = seed.acroeos.one:9876
p2p-peer-address = seed.eoseco.com:10021
p2p-peer-address = seed1.eos42.io:9876
p2p-peer-address = seed1.greymass.com:9876
p2p-peer-address = seed2.eos42.io:9876
p2p-peer-address = seed2.greymass.com:9876
p2p-peer-address = v.rito.su:9876
```

**Comment by jgiszczak** [comment]
What is the error reported immediately before `nodeos` terminates?  Connection closures do not cause `nodeos` to exit.

**Comment by quarksgluons** [comment]
There is no error message，so I didn't know why It shut down.
What I can do is just restart it or restart it from backup files.

**Comment by jgiszczak** [comment]
Without some error message, we have no way to diagnose why it shut down.

Check your system settings around RAM usage.  Many distributions of Linux are configured to kill processes that use large amounts of RAM.  Mainnet currently has 43.69 GiB of reserved RAM.  Not all of it is in use, but a good deal of it is.  A system or container with less than 8 GB available is unlikely to be able to run a mainnet-synched nodeos.

**Comment by quarksgluons** [comment]
The memory of my server is 8GB. Is this not enough?

**Comment by quarksgluons** [comment]
Generally, how high the configuration of the main network node needs to be synchronized. After one year, after two years, how much configuration is needed? I need to prepare.

**Comment by quarksgluons** [comment]
Ok, I will continue to observe, maybe It still will shut down.

The memory of my server is 8GB. Is this not enough?
Generally, what the configuration of the server needs to  synchronize main network node？
Then, after one year, two years, what configuration will needed? I need to know to prepare.

**Comment by jgiszczak** [comment]
A mainnet node without history_plugin today has a virtual size of over 6 GB.  Actual RAM used is a little under 2 GB.  With the history_plugin, it is considerably bigger.  In theory 8 GB is sufficient without history_plugin.  In practice, due to interactions with how boost shared_memory tries to reserve memory and interactions with how virtual machines manage memory, it may not be.

It's difficult to say what the required minimum will be in one or two years.  If EOS.IO is widely adopted and heavily used, you would expect 90% of the RAM available to be used.  The producers have enabled RAM expansion in the system contract, which is now making available 69.90GiB.  This number rises slowly every day.  How much of the available memory is used will depend on the adoption rate, which is unpredictable.

**Comment by quarksgluons** [comment]
Today, one of my nodes stopped again, but today I kept the system log. 
The log shows that it's memory occupancy rate is too high and the system killed it. And also need to replay when i restart it. I can only start from the backup.

Can this 90% memory use rate be reduced by some way or Some configuration, like 80% or 60% and the rest in virtual memory. This not only avoids this abnormal stop, but also reduces the memory requirements on the server.
```
...
Aug 25 15:03:59  xxx-xxx-xxx-xxx kernel: [2781980.456483] Out of memory: Kill process 10614 (nodeos) score 963 or sacrifice child
Aug 25 15:03:59 xxx-xxx-xxx-xxx kernel: [2781980.458006] Killed process 10614 (nodeos) total-vm:83923900kB, anon-rss:15795692kB, file-rss:0kB
...

```

**Comment by jgiszczak** [comment]
Use the mongo_db_plugin instead of the history_plugin.  History plugin stores everything in memory, and with a filter of "*", this is enormous, and will only get worse.  You will have to write your dapps to query mongo rather than nodeos RPC for history, but there are libraries available that make that easy.

**Comment by quarksgluons** [comment]
I've tried to synchronize the Mongo database, but it was so was that far from keeping up with the master node. So, I gave up that way.

I know that the way filter-on * will be memory problems, so I just add filtering to my account instead of filtering * now.

I have made a script on another server with 8 gigabytes of memory. The script stops for backing up and starts nodeos server again once a day . The nodeos server has been running steadily for more than a week. 

So , I feel that the daily restarting process can avoid nodeos takes up too much memory and being killed by linux. I don't know if I need to restart every 12 hours or even 1 hours in feture.
